lib/cn/kvset_split.c

```diff
@@ -0,0 +1,619 @@
+
+    struct hblock_builder *hbb_left;   /* hblock builder for the left hblock */
+    struct hblock_builder *hbb_right;  /* hblock builder for the right hblock */
+};
+
+/**
+ * kblock_copy_range() - copy all keys `k' in range start < k <= end from @kbd
+ *
+ * @kbd:      kblock descriptor
+ * @start:    start key (exclusive)
+ * @end :     end key (inclusive)
+ * @kblks_out: (output) blklist of kblocks where the keys are copied into
+ *
+ * Notes:
+ * (1) start == NULL implies a copy from the first key
+ * (2) end == NULL implies a copy until EOF
+ * (3) The output mblock in @kbid_out is not committed.
+ *     It is left to the caller to either commit or abort it.
+ * (4) This routine can generate > 1 output kblock due to some possible variability from key
+ *     compression
+ */
+static merr_t
+kblock_copy_range(
+    struct kblock_desc   *kbd,
+    const struct key_obj *start, /* exclusive */
+    const struct key_obj *end,   /* inclusive */
+    struct blk_list      *kblks_out)
+{
+    struct wbti *wbti = NULL;
+    struct kblock_builder *kbb;
+    struct key_obj cur = { 0 };
+    const void *kmd;
+    merr_t err;
+    uint64_t tot_keys = 0;
+
+    err = kbb_create(&kbb, kbd->cn, NULL);
+    if (err)
+        return err;
+
+    if (start) {
+        char kbuf[HSE_KVS_KEY_LEN_MAX];
+        struct kvs_ktuple seek;
+        uint klen;
+
+        key_obj_copy(kbuf, sizeof(kbuf), &klen, start);
+        kvs_ktuple_init_nohash(&seek, kbuf, klen);
+
+        err = wbti_create(&wbti, kbd->kd_mbd->map_base, kbd->kd_wbd, &seek, false, false);
+    } else {
+        err = wbti_create(&wbti, kbd->kd_mbd->map_base, kbd->kd_wbd, NULL, false, false);
+    }
+
+    while (!err && wbti_next(wbti, &cur.ko_sfx, &cur.ko_sfx_len, &kmd)) {
+        struct key_stats stats = { 0 };
+        size_t off = 0, kmd_cnt_off;
+        uint nvals;
+
+        wbti_prefix(wbti, &cur.ko_pfx, &cur.ko_pfx_len);
+
+        if (HSE_UNLIKELY(start && key_obj_cmp(&cur, start) == 0))
+            continue;
+        start = NULL;
+
+        if (HSE_UNLIKELY(end && key_obj_cmp(&cur, end) > 0))
+            break;
+
+        nvals = stats.nvals = kmd_count(kmd, &off);
+        kmd_cnt_off = off;
+
+        while (nvals--) {
+            struct kvs_vtuple_ref vref = { 0 };
+            u64 vseq;
+
+            /* Pass NULL for vgmap as vbidx is not used here */
+            wbt_read_kmd_vref(kmd, NULL, &off, &vseq, &vref);
+
+            switch (vref.vr_type) {
+            case vtype_val:
+            case vtype_cval:
+                stats.tot_vlen += (vref.vb.vr_complen ? : vref.vb.vr_len);
+                break;
+
+            case vtype_ival:
+                stats.tot_vlen += vref.vi.vr_len;
+                break;
+
+            case vtype_tomb:
+                ++stats.ntombs;
+                break;
+
+            case vtype_zval:
+                break;
+
+            case vtype_ptomb:
+            default:
+                assert(0);
```

Comment:
Since vtype_* are enums, I'd abort() on vtype_ptomb() and remove the default so that the compiler warns of unhandled cases, should someone add a new vtype_*.